js/src/vector/traits.ts

```diff
@@ -17,11 +17,26 @@
 
 import { Vector } from './vector';
 import { BoolVector } from './numeric';
-import * as Schema_ from '../format/Schema';
-import * as Message_ from '../format/Message';
-import Type = Schema_.org.apache.arrow.flatbuf.Type;
-import Field = Schema_.org.apache.arrow.flatbuf.Field;
-import FieldNode = Message_.org.apache.arrow.flatbuf.FieldNode;
+import { fb, FieldBuilder, FieldNodeBuilder } from '../format/arrow';
+
+export type Field = ( fb.Schema.Field | FieldBuilder );
+export type FieldNode = ( fb.Message.FieldNode | FieldNodeBuilder );
```

Comment:
This is great! I love it

Reply:
Hello.  This is probably not the best place to ask this question given that this is closed, if so, I can open a new jira ticket, just let me know what you prefer.  I'm looking to retrieve a json or csv dataset from an online web service, and then convert that to an arrow table.  The readJSON function in the json.ts file listed above seems like it would do the trick, but I don't see it available in the latest arrow release.  Can you point me to where that lives or how best to ingest a JSON (or CSV for that matter) file with the JS version of arrow and turn it into an arrow table?  I'm not readily finding examples for how to do this.

For what it's worth, here is a snippet of the ts code I'm working on for this:
`    
import { Table, RecordBatchWriter, RecordBatchStreamReader, RecordBatchFileReader, RecordBatchReader  } from "apache-arrow";
import Axios from 'axios';
let dwUrl = "https://www.nwfsc.noaa.gov/data/api/v1/source/trawl.operation_haul_fact/selection.json";
const response = await Axios.get(dwUrl);
const data = response.data;`

which can be found here:
https://github.com/nwfsc-fram/OceanTS/blob/master/src/utilities.ts

I want to be able to turn the response.data into an arrow table.

Thanks for all that you're doing with arrow!  It's a great effort and I'm really enjoying using it.

Reply:
Hi @toddhay-NOAA. This PR was related to implementing the JSON reader and writer for testing interop between the different language implementations (and has since been rewritten as the RecordBatchJSONReader/Writer classes). The JSON format isn't really suitable as an IPC mechanism in production setups, it's only meant for our integration tests.

It sounds like what you're really looking for is the Vector builders, which are presently a work-in-progress. I've been in the process of extracting what I have inside Graphistry to [this branch in my fork](https://github.com/trxcllnt/arrow/tree/js/data-builders/js). I've made a bit more progress on our internal versions and should be picking it up again early next week. If you need to get started sooner and/or want to collaborate feel free to shoot me an email and I can work to update this branch with my latest fixes.

You can start [here](https://github.com/trxcllnt/arrow/blob/c4aec7743286e6678bd0ea0c1247b9dcc79f3eb2/js/test/unit/builders/utils.ts#L86) to get an idea of what they'll look like. The idea is to enable aggregating chunks of Vectors into RecordBatches via iteration/async-iteration/streams. I do plan to add convenience APIs that expose the builders as Node/DOM streams similar to the Readers and Writers, but haven't gotten to that stage yet:

```ts
let fieldsAndChunks = [...getSourceColumns()].map(jsTypesToArrowBuilders);

let fields = fieldsAndChunks.map(({ field }) => field);
let columnChunks = fieldsAndChunks.map(({ chunks }) => chunks);
let recordBatches = []

for (let columns of zip(...chunks)) {
    recordBatches.push(RecordBatch.new(data, fields))
}

type ColumnStreamDesc = { colName: string, jsType: 'string', valuesStream: Iterable<any> };
function jsTypesToArrowBuilders({ colName, jsType, valuesStream }: ColumnStreamDesc) {
    let arrowType: DataType;
    let valuesThatIndicateNull: any[];
    switch (jsType) {
      case 'integer':
        arrowType = new Int32();
        valuesThatIndicateNull = [0x7ffffff];
        break;
      case 'float':
        arrowType = new Float64();
        valuesThatIndicateNull = [NaN];
        break;
      // etc.
      default:
        arrowType = new Utf8();
        valuesThatIndicateNull = [null, 'null', 'n/a', 'NA', ''];
        break;
    }
    let builder = DataBuilder.new(arrowType, valuesThatIndicateNull);
    return {
        field: new Field(colName, arrowType),
        chunks: builder.fromIterable(valuesStream)
    };
}
```

Reply:
Maybe open a JIRA about adding some documentation or making other improvements?

Reply:
@wesm ah for some reason I thought I'd already opened a JIRA for the builders. Adding one now.

Reply:
@toddhay-NOAA you can also look at this project for some examples of sending/receiving Arrow over HTTP via node: https://github.com/trxcllnt/fastify-arrow

If you want to inspect the JSON-encoded version of an existing Arrow table, you can reach into the internals and write out a Table in the JSON format like this:
```ts
// test.js
const { RecordBatchJSONWriter } = require('apache-arrow')
const table = Table.from(someDataSource())
RecordBatchJSONWriter.writeAll(table).pipe(process.stdout)
```

Then you can execute that script from the command line. The JSON isn't formatted, so I recommend piping it through [`jq`](https://stedolan.github.io/jq/) to fix up the indentation. The result should look something [like this](https://gist.github.com/trxcllnt/805aa7f066e9fda09a6827657f4eff75):
```
$ node test.js | jq
```

You can also pipe the JSON into the `arrow2csv` utility that we ship to inspect tables that way:
```
$ node test.js | jq -r '.' | npx arrow2csv | less
```

Reply:
@trxcllnt, thanks for your quick replies on all of this, it sounds Vector builders might be want I want.  I would like to get started this weekend if possible.  I see your [jira ](https://issues.apache.org/jira/browse/ARROW-5115) ticket for this and so will move the conversation over there.